File: stack/second_next_greater/second_next_greater.cpp
#include <vector>
#include <stack>
using namespace std;

vector<int> find_second_next_greater(vector<int> &nums) // LeetCode Q.2454.
{
    vector<int> second_next_greater(nums.size(), -1);

    stack<pair<int, int>> stack_1, stack_2; // Format: {num, idx}.
    vector<pair<int, int>> transporter;     // Format: {num, idx}.

    for (int idx = 0; idx < nums.size(); idx++)
    {
        while (!stack_2.empty() && stack_2.top().first < nums[idx])
        {
            int past_idx = stack_2.top().second;
            second_next_greater[past_idx] = nums[idx];
            stack_2.pop();
        }

        while (!stack_1.empty() && stack_1.top().first < nums[idx])
        {
            transporter.push_back(stack_1.top()); // Keep decreasing monotonicity.
            stack_1.pop();
        }

        while (!transporter.empty())
        {
            stack_2.push(transporter.back());
            transporter.pop_back();
        }

        stack_1.push({nums[idx], idx});
    }
    return second_next_greater;
}

File: stack/second_next_greater/second_next_greater.py

def find_second_next_greater(nums: list[int]) -> list[int]:  # LeetCode Q.2454.
    second_next_greater = [-1] * len(nums)
    stack_1, stack_2 = [], []  # Decreasing monotonic stacks: (num, idx).
    transporter = []  # Transport tuples from stack 1 to stack 2.

    for idx, num in enumerate(nums):
        while stack_2 and stack_2[-1][0] < num:
            _, past_idx = stack_2.pop(-1)
            second_next_greater[past_idx] = num

        while stack_1 and stack_1[-1][0] < num:
            transporter.append(stack_1.pop(-1))
        
        while transporter:
            stack_2.append(transporter.pop(-1))  # Ensure decreasing monotonicity.
        stack_1.append((num, idx))

    return second_next_greater
